File: an2_sem4/MPP/Homework/sapt56/java/service/src/main/java/ro/ubbcluj/cs/stefnmubb/festivalsellpoint/service/AbstractService.java
package ro.ubbcluj.cs.stefnmubb.festivalsellpoint.service;

import ro.ubbcluj.cs.stefnmubb.festivalsellpoint.domain.Entity;
import ro.ubbcluj.cs.stefnmubb.festivalsellpoint.repo.EntityRepoException;
import ro.ubbcluj.cs.stefnmubb.festivalsellpoint.repo.IRepo;

public abstract class AbstractService<ID, E extends Entity<ID>, R extends IRepo<ID, E>>
    implements IService<ID, E> {
    private final R repo;

    public R getRepo(){
        return repo;
    }

    public AbstractService(R repo) {
        this.repo = repo;
    }

    @Override
    public void add(E e) throws EntityRepoException {
        repo.add(e);
    }

    @Override
    public void update(E e) throws EntityRepoException {
        repo.update(e);
    }

    @Override
    public void remove(ID id) throws EntityRepoException {
        repo.remove(id);
    }

    @Override
    public E getById(ID id) throws EntityRepoException {
        return repo.getById(id);
    }

    @Override
    public Iterable<E> getAll() throws EntityRepoException {
        return repo.getAll();
    }


}

File: an2_sem4/MPP/Homework/sapt9/cs/FestivalSellpoint/FestivalSellpoint.Service/AbstractService.cs
using FestivalSellpoint.Domain;
using FestivalSellpoint.Repo;
using System.Collections.Generic;

namespace FestivalSellpoint.Service
{
    public class AbstractService<ID, E, R> : IService<ID, E> 
        where E : Entity<ID>
        where R : IRepo<ID, E>
    {
        protected R Repo { get; }

        public AbstractService(R repo)
        {
            Repo = repo;
        }

        public void Add(E e) => Repo.Add(e);        

        public IEnumerable<E> GetAll() => Repo.GetAll();

        public E GetById(ID id) => Repo.GetById(id);

        public void Remove(ID id) => Repo.Remove(id);

        public void Update(E e) => Repo.Update(e);        
    }
}
